

logger <- logger()

#### ---- Essential Genes ---- ####

info(logger, "Making a single data frame of gene essentiality.")

# the relevant data files are:
#    depmap19Q3.Achilles.common.essentials
#    depmap19Q3.common.essentials
#    depmap19Q3.nonessentials

essentiality_tib <- bind_rows(
        depmap19Q3.Achilles.common.essentials %>% add_column(label = "achilles_essential"),
        depmap19Q3.common.essentials %>% add_column(label = "common_essential"),
        depmap19Q3.nonessentials %>% add_column(label = "nonessential")
    ) %>%
    mutate(hugo_symbol = get_hugo_from_depmap_ids(gene),
           entrez_id = get_entrez_from_depmap_ids(gene)) %>%
    select(hugo_symbol, entrez_id, label)

if (nrow(essentiality_tib) > 0) {
    info(logger, paste("`essentiality_tib` has", nrow(essentiality_tib), "rows."))
} else {
    warn(logger, "`essentiality_tib` has 0 rows.")
}
info(logger, paste("Caching `essentiality_tib`."))
cache("essentiality_tib")


#### ---- CNA ---- ####

# CNA data comes from CCLE in a wide format: cell line X gene
# Gene level copy number data, log2 transformed with a pseudo count of 1.
# This is generated by mapping genes onto the segment level calls.

info("Preparing copy number information for the cell lines.")

ccle_copy_number <- depmap19Q3.CCLE.gene.cn %>%
    gather(key = "gene", value = "log_copy_number", -X) %>%
    mutate(hugo_symbol = get_hugo_from_depmap_ids(gene, "\\.")) %>%
    dplyr::rename(dep_map_id = "X") %>%
    select(dep_map_id, hugo_symbol, log_copy_number) %>%
    mutate(
        copy_number = 2**log_copy_number,
        copy_number_label = case_when(
            copy_number < 0.5 ~ "homo_del",
            copy_number < 1.5 ~ "het_del",
            copy_number > 2.5 ~ "amp",
            TRUE ~ "norm"
    ))

info(logger, paste0("`ccle_copy_number` has", nrow(ccle_copy_number), "rows."))

info(logger, "Caching `ccle_copy_number`.")
cache("ccle_copy_number")


#### ---- Mutations ---- ####

# The mutation data is pretty tidy. I cached two data tables, one with all of
# the mutations and one with only coding mutations.

info(logger, "Cleaning up CCLE mutation data.")

ccle_mutations <- depmap19Q3.CCLE.mutations %>%
    janitor::clean_names() %>%
    select(dep_map_id, hugo_symbol, entrez_gene_id, chromosome, start_position,
           end_position, variant_classification, variant_type, reference_allele,
           tumor_seq_allele1, codon_change, protein_change,
           is_deleterious, is_cosmi_chotspot) %>%
    dplyr::rename(is_cosmic_hotspot = "is_cosmi_chotspot")

ccle_mutations_coding <- ccle_mutations %>%
    filter(variant_classification %in% !!coding_mut_var_classes)

info(logger, "Caching `ccle_mutations`.")
cache("ccle_mutations")

info(logger, "Caching `ccle_mutations_coding`.")
cache("ccle_mutations_coding")


#### ---- RNA expression ---- ####

# RNA expression data comes from CCLE in a wide format: cell line X gene
# From the README:
#   RNAseq TPM gene expression data for just protein coding genes using RSEM.
#   Log2 transformed, using a pseudo-count of 1.

ccle_expression <- depmap19Q3.CCLE.expression %>%
    gather(key = "gene", value = "rna_expression", -X) %>%
    mutate(hugo_symbol = get_hugo_from_depmap_ids(gene, "\\.")) %>%
    dplyr::rename(dep_map_id = "X") %>%
    select(dep_map_id, hugo_symbol, rna_expression)

info(logger, "Caching `ccle_expression`.")
cache("ccle_expression")


#### ---- KRAS mutations ---- ####

info(logger, "Extracting KRAS mutants.")

kras_mutation_tib <- ccle_mutations_coding %>%
    filter(hugo_symbol == "KRAS") %>%
    mutate(
        codon = str_extract(protein_change, "[:digit:]+"),
        is_hotspot = codon %in% kras_hotspot_codons$char,
        allele = ifelse(
            is_hotspot,
            str_remove(protein_change, "p\\."),
            "other"
        )) %>%
    select(dep_map_id, codon, is_hotspot, allele, variant_classification)

kras_amps <- ccle_copy_number %>%
    filter(hugo_symbol == "KRAS" & copy_number_label == "amp") %>%
    select(dep_map_id, copy_number_label) %>%
    unique()

kras_mutation_tib <- full_join(kras_mutation_tib, kras_amps, by = "dep_map_id")

num_kras_cn <- sum(kras_mutation_tib$copy_number_label == "amp", na.rm = TRUE)
num_kras_muts <- n_distinct(kras_mutation_tib$dep_map_id)

info(logger, paste("There were", num_kras_cn, "cell lines with KRAS amplifications."))
info(logger, paste("There were", num_kras_muts, "cell lines with a KRAS alteration."))

info(logger, "Caching `kras_mutation_tib`.")
cache("kras_mutation_tib")

