
# Prepare the data from DepMap (2019 Q3 release)
info(logger, "Prepare the data from DepMap (2019 Q3 release)")

info(logger, "Making a single data frame of gene essentiality.")

# the relevant data files are:
#    depmap19Q3.Achilles.common.essentials
#    depmap19Q3.common.essentials
#    depmap19Q3.nonessentials

cache("essentiality_tib",
      depends = c("depmap19Q3.Achilles.common.essentials",
                  "depmap19Q3.common.essentials",
                  "depmap19Q3.nonessentials"),
{
    essentiality_tib <- bind_rows(
            depmap19Q3.Achilles.common.essentials %>% add_column(label = "achilles_essential"),
            depmap19Q3.common.essentials %>% add_column(label = "common_essential"),
            depmap19Q3.nonessentials %>% add_column(label = "nonessential")
        ) %>%
        mutate(hugo_symbol = get_hugo_from_depmap_ids(gene),
               entrez_id = get_entrez_from_depmap_ids(gene)) %>%
        select(hugo_symbol, entrez_id, label)

    log_rows(logger, essentiality_tib, "essentiality_tib")
    info(logger, paste("Caching `essentiality_tib`."))

    return(essentiality_tib)
})


#### ---- CNA ---- ####

# CNA data comes from CCLE in a wide format: cell line X gene
# Gene level copy number data, log2 transformed with a pseudo count of 1.
# This is generated by mapping genes onto the segment level calls.

cache('ccle_copy_number', depends = "depmap19Q3.CCLE.gene.cn",
{
    info(logger, "Preparing copy number information for the cell lines.")

    ccle_copy_number <- depmap19Q3.CCLE.gene.cn %>%
        gather(key = "gene", value = "log_copy_number", -X) %>%
        mutate(hugo_symbol = get_hugo_from_depmap_ids(gene, "\\.")) %>%
        dplyr::rename(dep_map_id = "X") %>%
        select(dep_map_id, hugo_symbol, log_copy_number) %>%
        mutate(
            copy_number = 2**log_copy_number,
            copy_number_label = case_when(
                copy_number < 0.5 ~ "homo_del",
                copy_number < 1.5 ~ "het_del",
                copy_number > 2.5 ~ "amp",
                TRUE ~ "norm"
        ))

    log_rows(logger, ccle_copy_number, "ccle_copy_number")
    info(logger, "Caching `ccle_copy_number`.")
    return(ccle_copy_number)
})




#### ---- Mutations ---- ####

# The mutation data is pretty tidy. I cached two data tables, one with all of
# the mutations and one with only coding mutations.

cache("ccle_mutations", depends = "depmap19Q3.CCLE.mutations",
{
    info(logger, "Cleaning up CCLE mutation data.")

    ccle_mutations <- depmap19Q3.CCLE.mutations %>%
        janitor::clean_names() %>%
        select(dep_map_id, hugo_symbol, entrez_gene_id, chromosome, start_position,
               end_position, variant_classification, variant_type, reference_allele,
               tumor_seq_allele1, codon_change, protein_change,
               is_deleterious, is_cosmi_chotspot) %>%
        dplyr::rename(is_cosmic_hotspot = "is_cosmi_chotspot")

    log_rows(logger, ccle_mutations, "ccle_mutations")
    info(logger, "Caching `ccle_mutations`.")
    return(ccle_mutations)
})


cache("ccle_mutations_coding", depends = "ccle_mutations",
{
    ccle_mutations_coding <- ccle_mutations %>%
        filter(variant_classification %in% !!coding_mut_var_classes)
    log_rows(logger, ccle_mutations_coding, "ccle_mutations_coding")
    info(logger, "Caching `ccle_mutations_coding`.")
    return(ccle_mutations_coding)
})


#### ---- RNA expression ---- ####

# RNA expression data comes from CCLE in a wide format: cell line X gene
# From the README:
#   RNAseq TPM gene expression data for just protein coding genes using RSEM.
#   Log2 transformed, using a pseudo-count of 1.

cache("ccle_expression", depends = "depmap19Q3.CCLE.expression",
{
    ccle_expression <- depmap19Q3.CCLE.expression %>%
        gather(key = "gene", value = "rna_expression", -X) %>%
        mutate(hugo_symbol = get_hugo_from_depmap_ids(gene, "\\.")) %>%
        dplyr::rename(dep_map_id = "X") %>%
        select(dep_map_id, hugo_symbol, rna_expression)

    log_rows(logger, ccle_expression, "ccle_expression")
    info(logger, "Caching `ccle_expression`.")
    return(ccle_expression)
})




#### ---- KRAS mutations ---- ####


cache("kras_mutation_tib",
      depends = c("ccle_mutations_coding", "ccle_copy_number"),
{
    info(logger, "Extracting KRAS mutants.")

    kras_mutation_tib <- ccle_mutations_coding %>%
        filter(hugo_symbol == "KRAS") %>%
        mutate(
            codon = str_extract(protein_change, "[:digit:]+"),
            is_hotspot = codon %in% kras_hotspot_codons$char,
            allele = ifelse(
                is_hotspot,
                str_remove(protein_change, "p\\."),
                "other"
            )) %>%
        select(dep_map_id, codon, is_hotspot, allele, variant_classification)

    kras_amps <- ccle_copy_number %>%
        filter(hugo_symbol == "KRAS" & copy_number_label == "amp") %>%
        select(dep_map_id, copy_number_label) %>%
        unique()

    kras_mutation_tib <- full_join(kras_mutation_tib, kras_amps, by = "dep_map_id")

    log_rows(logger, kras_mutation_tib, "kras_mutation_tib")

    num_kras_cn <- sum(kras_mutation_tib$copy_number_label == "amp", na.rm = TRUE)
    num_kras_muts <- n_distinct(kras_mutation_tib$dep_map_id)

    info(logger, paste("There were", num_kras_cn, "cell lines with KRAS amplifications."))
    info(logger, paste("There were", num_kras_muts, "cell lines with a KRAS alteration."))

    info(logger, "Caching `kras_mutation_tib`.")
    return(kras_mutation_tib)
})





#### ---- Sample Info ---- ####

cache("cell_lines",
      depends = c("depmap19Q3.sample.info", "kras_mutation_tib"),
{
    info(logger, "Preparing sample (cell line) information.")

    cell_lines <- depmap19Q3.sample.info %>%
        janitor::clean_names() %>%
        select(dep_map_id, stripped_cell_line_name, sex, age,
               lineage, lineage_subtype, lineage_sub_subtype,
               disease, disease_subtype,
               sample_collection_site, primary_or_metastasis) %>%
        mutate(sex = ifelse(sex == "", NA, sex)) %>%
        mutate(cancer = case_when(
            disease == "Colon/Colorectal Cancer" ~ "COAD",
            disease == "Pancreatic Cancer" ~ "PAAD",
            disease == "Myeloma" ~ "MM",
            lineage_subtype == "lung_NSC" ~ "LUAD"
        )) %>%
        left_join(kras_mutation_tib, by = "dep_map_id")

    log_rows(logger, cell_lines, "cell_lines")
    info(logger, paste("`cell_lines` has", n_distinct(cell_lines$dep_map_id), "cells."))
    info(logger, "Caching `cell_lines.`")
    return(cell_lines)
})


#### ---- Gene Effect ---- ####
# the effect of KOing a gene

cache("gene_effect", depends = "depmap19Q3.Achilles.gene.effect",
{
    info(logger, "Preparing the gene effect data.")

    gene_effect <- depmap19Q3.Achilles.gene.effect %>%
        gather(key = "gene", value = "gene_effect", -X) %>%
        mutate(hugo_symbol = get_hugo_from_depmap_ids(gene, "\\.\\."),
               hugo_symbol = str_replace_all(hugo_symbol, "\\.", "-")) %>%
        dplyr::rename(dep_map_id = "X") %>%
        select(dep_map_id, hugo_symbol, gene_effect)

    log_rows(logger, gene_effect, "gene_effect")
    info(logger, "Caching `gene_effect`.")
    return(gene_effect)
})


#### ---- Dependency Liklihood ---- ####
# the probability that the cell line is dependent on the target gene

cache("dependency_prob", depends = "depmap19Q3.Achilles.gene.dependency",
{
    info(logger, "Preparing the gene dependency probability data.")

    dependency_prob <- depmap19Q3.Achilles.gene.dependency %>%
        gather(key = "gene", value = "prob", -X) %>%
        mutate(hugo_symbol = get_hugo_from_depmap_ids(gene, "\\.\\."),
               hugo_symbol = str_replace_all(hugo_symbol, "\\.", "-")) %>%
        dplyr::rename(dep_map_id = "X") %>%
        select(dep_map_id, hugo_symbol, prob)

    log_rows(logger, dependency_prob, "dependency_prob")
    info(logger, "Caching `dependency_prob`.")
    return(dependency_prob)
})



#### ---- RNAi ---- ####

# The Achilles project began by using an RNAi screen

cache("rnai_effect",
      depends = c("depmap19Q3.sample.info",
                  "depmap19Q3.D2.combined.gene.dep.scores"),
{
    info(logger, "Preparing the RNAi screen data.")

    ccle_to_depmap_names <- depmap19Q3.sample.info %>%
        janitor::clean_names() %>%
        select(dep_map_id, ccle_name) %>%
        unique()

    rnai_effect <- depmap19Q3.D2.combined.gene.dep.scores %>%
        gather(key = "ccle_name", value = "rnai_effect", -X) %>%
        mutate(
            hugo_symbol = get_hugo_from_depmap_ids(X),
            ccle_name = str_remove(ccle_name, "^X")) %>%
        left_join(ccle_to_depmap_names, by = "ccle_name") %>%
        select(dep_map_id, hugo_symbol, rnai_effect)

    log_rows(logger, rnai_effect, "rnai_effect")
    info(logger, "Caching `rnai_effect`.")
    return(rnai_effect)
})
