---
title: '2019-10-05'
author: ~
date: '2019-10-05'
slug: '2019-10-05'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2019-10-05'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---

## Modeling KRAS allele-specific synthetic lethality (cont.)

Continuing from yesterday, I worked on plotting the box-plots of genes found to be significantly depleted in *KRAS* mutants.
There was a bug in 'ggpubr' that installing the dev. version from GitHub fixed.
I also ignored cell lines of LUAD with *KRAS* G13D mutations because they are irrelevant to patients.

Depletion box-plots were made for the genes that passed the following criteria:

1. The linear model of depletion effect and RNA expression was non-significant (p-value < 0.01).
2. The ANOVA across all alleles was significant at a relatively stringent cut-off (p-value < 0.01).
3. There was at least one significant pairwise comparison (FDR adj. p-value < 0.05)

All of the box-plots are in "graphs/10_10_linear-modeling-syn-let_boxplots/".
Below are some of the more interesting genes to pop out of the analysis.

### COAD

<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/COAD-ARID5B.svg" width = 400 />

From [UniProt ARI5B_HUMAN](https://www.uniprot.org/uniprot/Q14865):

> Transcription coactivator that binds to the 5'-AATA[CT]-3' core sequence and plays a key role in adipogenesis and liver development. 
> Acts by forming a complex with phosphorylated PHF2, which mediates demethylation at Lys-336, leading to target the PHF2-ARID5B complex to target promoters, where PHF2 mediates demethylation of dimethylated 'Lys-9' of histone H3 (H3K9me2), followed by transcription activation of target genes...
> Required for adipogenesis: regulates triglyceride metabolism in adipocytes by regulating expression of adipogenic genes.
> Overexpression leads to induction of smooth muscle marker genes, suggesting that it may also act as a regulator of smooth muscle cell differentiation and proliferation.

<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/COAD-EIF3M.svg" width = 400 />
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/COAD-GAS6.svg" width = 400 />
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/COAD-SRP9.svg" width = 400 />

From [UniProt SRP09_HUMAN](https://www.uniprot.org/uniprot/P49458):

> Signal-recognition-particle assembly has a crucial role in targeting secretory proteins to the rough endoplasmic reticulum membrane. 
> SRP9 together with SRP14 and the Alu portion of the SRP RNA, constitutes the elongation arrest domain of SRP. 
> The complex of SRP9 and SRP14 is required for SRP RNA binding.

<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/COAD-UNK.svg" width = 400 />

---

### LUAD

<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-CCDC107.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-CCDC113.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-DHFR.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-FAAP100.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-GPN2.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-HDAC1.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-HSBP1.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-MAPK6.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-SCAF4.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-STAMBP.svg" width = 400 >

From [UniProt: STABP_HUMAN](https://www.uniprot.org/uniprot/O95630):

> Zinc metalloprotease that specifically cleaves 'Lys-63'-linked polyubiquitin chains.
> Does not cleave 'Lys-48'-linked polyubiquitin chains (By similarity). 
> Plays a role in signal transduction for cell growth and MYC induction mediated by IL-2 and GM-CSF.
> Potentiates BMP (bone morphogenetic protein) signaling by antagonizing the inhibitory action of SMAD6 and SMAD7. 
> Has a key role in regulation of cell surface receptor-mediated endocytosis and ubiquitin-dependent sorting of receptors to lysosomes. 
> Endosomal localization of STAMBP is required for efficient EGFR degradation but not for its internalization (By similarity). 
> Involved in the negative regulation of PI3K-AKT-mTOR and RAS-MAP signaling pathways.

<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-TP53.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/LUAD-TP53BP1.svg" width = 400 >

---

### PAAD

<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/PAAD-CCT8.svg" width = 400 >
<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/PAAD-EGLN2.svg" width = 400 >
From [UniProt: EGLN2_HUMAN](https://www.uniprot.org/uniprot/Q96KS0):

> Cellular oxygen sensor that catalyzes, under normoxic conditions, the post-translational formation of 4-hydroxyproline in hypoxia-inducible factor (HIF) alpha proteins. 
> Hydroxylates a specific proline found in each of the oxygen-dependent degradation (ODD) domains (N-terminal, NODD, and C-terminal, CODD) of HIF1A. 
> Also hydroxylates HIF2A...
> Hydroxylated HIFs are then targeted for proteasomal degradation via the von Hippel-Lindau ubiquitination complex. 
> Under hypoxic conditions, the hydroxylation reaction is attenuated allowing HIFs to escape degradation resulting in their translocation to the nucleus, heterodimerization with HIF1B, and increased expression of hypoxy-inducible genes. 
> EGLN2 is involved in regulating hypoxia tolerance and apoptosis in cardiac and skeletal muscle. 
> Also regulates susceptibility to normoxic oxidative neuronal death. 
> Links oxygen sensing to cell cycle and primary cilia formation by hydroxylating the critical centrosome component CEP192 which promotes its ubiquitination and subsequent proteasomal degradation. 
> Hydroxylates IKBKB, mediating NF-kappaB activation in hypoxic conditions. 
> Target proteins are preferentially recognized via a LXXLAP motif.

<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/PAAD-JUN.svg" width = 400 >

From [UniProt JUN_HUMAN](https://www.uniprot.org/uniprot/P05412):

> Transcription factor that recognizes and binds to the enhancer heptamer motif 5'-TGA[CG]TCA-3' (PubMed:10995748, PubMed:22083952). 
> Promotes activity of NR5A1 when phosphorylated by HIPK3 leading to increased steroidogenic gene expression upon cAMP signaling pathway stimulation (PubMed:17210646). 
> Involved in activated KRAS-mediated transcriptional activation of USP28 in colorectal cancer (CRC) cells (PubMed:24623306). 
> Binds to the USP28 promoter in colorectal cancer (CRC) cells (PubMed:24623306).


--- 

#### Notes for tomorrow

There were some genes that have way too many data points (see example below).
I need to figure out what is causing this.
My guess is that they are related genes and the gene names are being truncated by `get_hugo_from_depmap_ids()`.

<img src="/img/graphs/10_10_linear-modeling-syn-let_boxplots/COAD-KRTAP2.svg" width = 400>

