---
title: '2019-10-25'
author: ~
date: '2019-10-25'
slug: '2019-10-25'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2019-10-25'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---


## Meeting with GM

I didn't have any time to work on this project today. 
I did meet with GM to go over a few things about his scripts and the figures,
It went well and he seemed pleased with the current status and my proposed changes.

Because I couldn't do any coding today, I took some time in the evening to brainstorm for the project.
My thoughts are below.


## Brainstorming

### Fixing the last panel of Figure 2

One thing I want to change is the end of Figure 2.
GM told me that the y-axis is not strictly the probability of getting the mutation.
This is what Kevin, Peter, and I thought that it represented, but it doesn't according to GM.
*We want to show the observed frequency of the alleles compared to their predicted frequency based on the trinucleotide context.*
The reader should look at this plot, read the axes as "observed vs. predicted" and immediately have a strong understanding concept of what is shown.
There is a paper that previously looked at this for the *KRAS* alleles; I should go see how they did the calculation.


### Logisitic regression for synthetic lethality

I began working on logistic regression for the synthetic lethality - regressing on the *KRAS* allele with the gene effect as the predictor.
I am still working on it, and it seems like a logical analysis to attempt.
According to my notes from yesterday, my next attempt is to run a logistic regression on each allele vs. the rest.

In addition, I could try a logistic LASSO with a logistic regression on the *KRAS* alleles using *all* of the alleles as predictors.
From a quick Google search, this seems like a good place to start: [*Penalized Logistic Regression Essentials in R: Ridge, Lasso and Elastic Net*](http://www.sthda.com/english/articles/36-classification-methods-essentials/149-penalized-logistic-regression-essentials-in-r-ridge-lasso-and-elastic-net/).


### Comparing overall dependencies of cell lines by *KRAS* alleles

GM had a good idea during our meeting (credit where credit is due). 
He suggested looking at the overall distribution of lethal genes in the cell lines of each *KRAS* allele.
I am picturing a ['ggridges'](https://cran.r-project.org/web/packages/ggridges/index.html) plot, one for each allele.
The thought was that perhaps "stronger" alleles will have fewer dependencies.
The tricky part will be figuring out which genes to include.
I should start by including all of the genes, though it is unlikely to show a difference between any of the alleles.
I can also use information from the essential and nonessential genes.
I would be interesting to just see the distributions for the "pan-essential" genes.

On a similar note, what about histograms for the genes in the KEGG pathways, CGC genes, and BioID genes?
Normalize within a gene, and then show 'ggridges' for the genes in each of those.
Then, underneath, a heatmap of the most variable genes in the gene set showing their values in the cell lines.
(I included a quick sketch in my notebook relating this thought.)
I could also do this for any gene sets found to be enriched in the genes in the whole-cancer heatmap (not within individual clusters - they are too small), or with the gene sets found to be enriched in GSEA.


### A term for reduced comutation but not mutual exclusivity

**We need a different term than "mutually exclusive" when referring to genes that frequently mutate with a *KRAS* allele, just less often than would be predicted by random.**
The term "mutually exclusive" should be saved for *BRAF* and *EGFR*-like interactions, but not for *TP53*.
However, I do not want to call it *reduced* comutation.
(GM mentioned that I had used a good term a little while ago, but couldn't recall it, I should look in the manuscript to see if it is there.)
This is essential for the believability of the paper.

For plotting, we will want to come up with a heurisitc for separating between these three classes.
Comutation and the others are easy, but between *BRAF*-like interactions and *TP53*-like interactions will be a bit trickier.


### Comutation networks

I need to use more stringent cutoffs for p-value, genetic events, and overall frequency of mutation before continuing further.
I can also include supplemental plots for the distribution and cumulative distribution of these metrics.

Jake Lee (Park Lab) brought up the good idea of representing the comutation and reduced comutation genes using a canonical rainfall plot.
My hesitation is because a lot of these events occur at low frequencies, the plot will be very sparse and unimpressive.
For example, the *PORCN* comutation interaction with *A146T* in COAD is interesting because it is mutated in 3 or 4 of the A146T samples and about the same number of other samples. 
Just given the first value, the interaction seems uninteresting; the second value provides context.
The ratio of these two values is the odds ratio, so perhaps somehow using that will increase the potency.
Maybe a rainfall plot with the color representing the OR?
If I can figure out a good plot for these interactions, then I can dive further into a few in particular to make some hypotheses (eg. what are the *PORCN* mutations?, are these likely disruptive?, are they located on the same  part of the protein?, etc.)

For the rainfall plots, one tricky part will be deciding what genes to include.
A few options are:

1. For each allele, the genes with the most number of interactions.
2. For each allele, the genes with the highest OR.
3. Enriched gene sets across all of the hits.
4. *A priori* selected gene sets.

---

I think this should give me plenty to work on, for now.
