---
title: '2019-10-08'
author: ~
date: '2019-10-08'
slug: '2019-10-08'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2019-10-08'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---


## Plan for analysis

I did not have time to work on this project today, but I found time in the evening to think about how I want to progress.

### RC-test results analysis

Tomorrow, I will focus on moving forward with analyzing the results from the RC-test.
I need to merge in the number of mutation events in order to remove the genetic interactions that do not meet a low standard.
For comutation, there should be at least 2 co-mutation events.
For mutual exclusivity, the other gene should be mutated in at least 0.01% of all the cancer's samples.
This will remove a lot of mutually exclusive events that are caused by the mutation just being very rare.

I will then compare the results to those from the Fisher's exact test:

1. How many genes are shared by each allele between the two analyses (Venn diagram for each allele)
2. For the shared genes, how do the p-values compare (scatter plot with x and y-axis as p-value from each analysis; a line for $y = x$)

Finally, I will look for interesting results in the significant genes.
This will start with highlight CGC genes.
Also those found from the BioID screen of *KRAS* and *KRAS* mutants.
Lastly, I will use WebGestalt, EnrichR, or PANTHER (PMID: 23868073 for updated overview) to find functional patterns in the gene lists.

For the paper, I see the text reading something to the effect of, "X tended to comutate with *KRAS* A and B but no interaction was detected for *KRAS* C nor D," for individual CGC genes. Also, for any enriched biological processes, the text would look something like, "There was an enrichment for function Y in the genes found to comutate with *KRAS* E, potentially suggesting ..."

Here is where I will stop until completing the synthetic lethal analysis.
Once I have completed that to a satisfactory degree, I will look for overlap between the hits from each analysis.


### Allele-specific synthetic lethality

So far, I have only completed the simplest possible analysis to find allele-specific synthetic lethality.
Briefly, I removed the genes that could be explained by their RNA expression.
Then, I conducted an ANOVA and pair-wise comparisons to find differences in lethality between cells with the various *KRAS* alleles, including WT.
Some of the most interesting results can be found on yesterday's notebook, and all plots are available in "graphs/10_10_linear-modeling-syn-let_bosplots/".

I want a way of visualizing all of these results (especially for the meeting with KH and PJP).
The obvious way is to produce a heatmap of patient vs. gene with the genes that were found to have a significant difference.
Then, I could show specific box-plots for genes of interest (CGC, *KRAS* interactors from BioID screen, specific KEGG pathways).

Another approach could be to look for enriched biological processes/pathways.
To this end, we could employ an analysis similar to GSEA.
We could run GSEA as one *KRAS* allele against all other cell lines or just the *KRAS* WT cell lines.
I will also look into GSVA and ssGSEA, two popular variants of GSEA.

A quick analysis that could be very fruitful is to conduct a PCA on the gene-space and see if the cell lines cluster by *KRAS* allele.
If they do, that would lend credence to our claim that the *KRAS* allele is important.
(It may be useful to plot the first few PCs using tSNE or UMAP.)
If there appear to be clusters, then some sort of clustering (K-means, hierarchical clustering, DBSCAN, spectral clustering, etc.) could be used to make the claim concrete.
Even just a difference between KRAS and WT would be supportive of our claim.

LASSO of KRAS WT vs mutant starting with all genes as covariates.
Use the genes with non-trivial parameters in a random forest classifier or niave bayes classifier to predict the *KRAS* allele (including WT).
