---
title: '2019-11-01'
author: ~
date: '2019-11-01'
slug: '2019-11-01'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2019-11-01'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---


## Descriptive plots for cell lines used by DepMap (cont.)

I improved the mutation tile plots from yesterday by adding bar-plots for the margins.
The images are available in yesterday's notebook.

## Overall dependencies of the cell lines by *KRAS* allele

GM had a good suggesting of looking at the overall distribution of dependencies of the cell lines by *KRAS* allele.
Perhaps, the "stronger" alleles will have reduced dependencies overall.

The analysis was done in "src/10_45_overall-dependencies-by-allele.R"

I made two plots for each cancer and see nothing from either of them.
The left plot shows the distribution of dependencies for the alleles (each gene's score was averaged across the represented cell lines.)
The right plot shows the quantile values for the gene scores of each allele.

![](/img/graphs/10_45_overall-dependencies-by-allele/dependency-overview_COAD.svg)

![](/img/graphs/10_45_overall-dependencies-by-allele/dependency-overview_LUAD.svg)

![](/img/graphs/10_45_overall-dependencies-by-allele/dependency-overview_PAAD.svg)

(I also tried scaling the gene scores across all cell lines before average within each allele, but found nothing interesting, still.)

## Specify therapeutically viable genes from synthetic lethal hits

The heatmaps showing the allele-specific dependencies show any gene with a detectable difference across the alleles.
I want to specify which hits are "therapeutically viable".
To qualify for this, the gene was be lethal in one context, but not in another.
Therefore, of the hits, I applied the following filters:

- in at least one, but not all, alleles, the average effect was stronger than the cut-off derived from the non-essential and essential genes
- the difference is statistically significant

This analysis was done in "src/10_16_linear-modeling-syn-let_therapeutics.R"

During the analysis, I noticed that some data from cell lines was repeated.
I looked back and found that that was used for the original analysis.
I have gone back to "src/10_10_linear-modeling-syn-let.R" and fixed the problem.
**TODO:** I still need to fix one bug in plotting of the results from the fixed analysis.

---

## Notes on observed vs. predicted *KRAS* allele frequency

In "Assessment of mutation probabilities of KRAS G12 missense mutants and their long-timescale dynamics by atomistic molecular simulations and Markov state modeling" Pantsar *et al.* (2018, PMID: [30199525](https://www.ncbi.nlm.nih.gov/pubmed/?term=30199525)) conducted an analysis of the frequency of *KRAS* alleles in different cancers, measuring the extent to which the alleles were distributed at random.
They found that the alleles could *not* be explained by random chance, especially G12R in PAAD.
Below is their abstract:

> A mutated KRAS protein is frequently observed in human cancers.
> Traditionally, the oncogenic properties of KRAS missense mutants at position 12 (G12X) have been considered as equal.
> Here, by assessing the probabilities of occurrence of all KRAS G12X mutations and KRAS dynamics we show that this assumption does not hold true.
> Instead, our findings revealed an outstanding mutational bias.
> We conducted a thorough mutational analysis of KRAS G12X mutations and assessed to what extent the observed mutation frequencies follow a random distribution.
> Unique tissue-specific frequencies are displayed with specific mutations, especially with G12R, which cannot be explained by random probabilities.
> To clarify the underlying causes for the nonrandom probabilities, we conducted extensive atomistic molecular dynamics simulations (170 μs) to study the differences of G12X mutations on a molecular level.
> The simulations revealed an allosteric hydrophobic signaling network in KRAS, and that protein dynamics is altered among the G12X mutants and as such differs from the wild-type and is mutation-specific.
> The shift in long-timescale conformational dynamics was confirmed with Markov state modeling.
> A G12X mutation was found to modify KRAS dynamics in an allosteric way, which is especially manifested in the switch regions that are responsible for the effector protein binding.
> The findings provide a basis to understand better the oncogenic properties of KRAS G12X mutants and the consequences of the observed nonrandom frequencies of specific G12X mutations.

Below is an excerpt from the analysis of mutation frequencies.
They used the ratio of transition to transversions to create their null distribution.

> We evaluated how random the occurrences of the specific G12X mutations are.
> To this end, we used the transition:transversion mutation ratio as a figure of merit, and compared this figure to a value of 2.3, which is the ratio for missense mutations observed in large-scale genomic analyses (34,35).
> If the mutations would take place randomly, G12D and G12S mutations should be the most abundant mutations as they are transition mutations (S2 Fig).
> G12D mutation is consistent with this view, as it occurs very often in all tissues.
> Meanwhile, G12S is not consistent with this behavior at all, as it occurs in tumors, perhaps surprisingly, very rarely.
> Also, regardless of the tissue type, the G12V mutation is over-expressed compared to values expected based on the assumption of random occurrences.
> Concluding, the mutations’ probabilities of occurrences are not consistent with a transition:transversion mutation ratio based on a random process.

**Supp. Fig. 2**

![](https://journals.plos.org/ploscompbiol/article/file?id=10.1371/journal.pcbi.1006458.s002&type=supplementary)

### New metric

We want to plot observed vs. predicted frequency of the *KRAS* mutations; we need a new metric for predicted.
The null is that the *KRAS* allele that each individual has was decided at random - the choice of *KRAS* allele was random and any would suffice.
I need to figure out some measurement that calculates the predicted frequency of each allele given this null.

--- 

## Notes from meeting with KH

- remove *EGFR*-activated mutants from *KRAS* WT in DepMap.

--- 

## Comments from Jake about oncoplots

- Overall, he was very positive about them.
- Look more closely at the overlap of the other genes found to comutate/mut. ex. with a *KRAS* allele and see if any of them lie in the same pathways.
- Limit the number of rows shown to just those with obviously striking results.
