---
title: '2019-11-11'
author: ~
date: '2019-11-11'
slug: '2019-11-11'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2019-11-11'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---


## Refine rainfall plots

I first had to fix a mistake in the creation of the MAF-compliant tibble by removing hypermutants.
This had no effects on the results because the hypermutants were removed from the initial tests for comutation and mutual exclusivity.
The only effect was to make some comutation interactions look even more striking.

The second fix was to set some custom sizes for the manually-selected plots.
This was purely aesthetic.

---

## Coloration of synthetic lethal heat maps

I updated the color scale used for the heatmaps to use the orange and purple for up and down effects, respectively:

<img src="/img/graphs/00_miscellaneous/genetic_interaction_pal.svg" width=400 />

This was pretty simple to do, and for future reference, here was all I had to do.

```r
pal <- c(synthetic_lethal_pal["down"], "grey95", synthetic_lethal_pal["up"])
pal <- colorRampPalette(pal)(7)

ph <- pheatmap::pheatmap(
    mod_data,
    color = pal,
    ...
)
```

## Fix observed vs. predicted allele frequency plots

Originally, GM used some funky formulation to create the observed vs. predicted plots was way to convoluted - he wasn't even able to explain it to me in person, so how could I do it in text?
Therefore, I wanted to create a simpler formulation to answer the question: **If a sample was to have a KRAS mutation, what is the most likely one for it to get if decided by mutational signatures?**

My formulation is as follows:

- for a tumor sample:
    + for each trinucleotide mutation $m$ with context $c$ that can cause a KRAS mutation:
        * calculate: $\frac{|m|}{|c|}$
    + for each allele $a$:
        * $p(a) = \sum{p(m)} \quad \text{where} \quad m \rightarrow a$
    + normalize such that $\{A\} \Rightarrow \sum^{A}{a} = 1$

This approach assumes that all codons are the same.
If we were interested in only G12 mutations (the null hypothesis being that all G12 mutations are equal), then the same approach is followed again, but only normalizing the G12 mutations at the end.

The data I used was from GM:

- "/n/data1/hms/dbmi/park/gmelloni/cooc_mutex_driver/data/cbioportal/":
    - "paad_luad_coadread_skcm_mm_tricLong_newAug2019.rds"
    - "context_and_signa_paad_luad_coadread_skcm_mm_NEW.rds"

It was prepared in "munge/50_trinucleotide-mutations.R"

The analysis was conducted in "src/50_10_observed-predicted-kras-alleles.R".
The box plot below shows the distribution of predicted allele frequencies in each cancer.

![](/img/graphs/50_10_observed-predicted-kras-alleles/predicted_kras_allele_frequency_boxplot.svg)

The bar plot shows the same data, just taking the average across all samples.
The error bars indicate the SEM.

![](/img/graphs/50_10_observed-predicted-kras-alleles/predict_kras_allele_frequency_barplot1.svg)

The following bar plot, again, shows the same results, just stratified by allele across cancers.

![](/img/graphs/50_10_observed-predicted-kras-alleles/predict_kras_allele_frequency_barplot2.svg)

Below are the results for two different null hypothesis: 1) all *KRAS* hotspots are equivalent, and 2) all *KRAS* G12 alleles are equivalent.

### COAD

<img src="/img/graphs/50_10_observed-predicted-kras-alleles/obs_pred_plot_stats_COAD.svg" width=400 />

<img src="/img/graphs/50_10_observed-predicted-kras-alleles/obs_pred_plot_g12_stats_COAD.svg" width=400 />

### LUAD

<img src="/img/graphs/50_10_observed-predicted-kras-alleles/obs_pred_plot_stats_LUAD.svg" width=400 />

<img src="/img/graphs/50_10_observed-predicted-kras-alleles/obs_pred_plot_g12_stats_LUAD.svg" width=400 />

### MM

<img src="/img/graphs/50_10_observed-predicted-kras-alleles/obs_pred_plot_stats_MM.svg" width=400 />

<img src="/img/graphs/50_10_observed-predicted-kras-alleles/obs_pred_plot_g12_stats_MM.svg" width=400 />

### PAAD

<img src="/img/graphs/50_10_observed-predicted-kras-alleles/obs_pred_plot_stats_PAAD.svg" width=400 />

<img src="/img/graphs/50_10_observed-predicted-kras-alleles/obs_pred_plot_g12_stats_PAAD.svg" width=400 />



I ran a binomial test to find which observed rates of the alleles matched the predicted.
The alleles that could be predicted (p > 0.05) trinucleotide context are shown below.

| Cancer | Allele | Obs. allele freq. | Pred. allele freq | *p*-value |
|:------:|:------:|:-----------------:|:-----------------:|:---------:|
| COAD   | G12A   | 0.0371            | 0.0314            | 0.445     |
| COAD   | G12R   | 0.00977           | 0.0171            | 0.301     |
| COAD   | Q61L   | 0.00977           | 0.0133            | 0.697     |
| LUAD   | Q61L   | 0.0141            | 0.0249            | 0.503     |
| MM     | G12A   | 0.0604            | 0.0431            | 0.171     |
| MM     | G12C   | 0.0302            | 0.0332            | 1         |
| MM     | G12D   | 0.121             | 0.0910            | 0.108     |
| MM     | G12V   | 0.0717            | 0.0602            | 0.436     |
| MM     | G13D   | 0.128             | 0.104             | 0.191     |
| MM     | Q61R   | 0.0528            | 0.0453            | 0.552     |

The alleles that could be predicted (p > 0.05) when only considering the G12 alleles are show below.

| Cancer | Allele | Obs. allele freq. | Pred. allele freq | *p*-value |
|:------:|:------:|:-----------------:|:-----------------:|:---------:|
| COAD   | G12D   | 0.293             | 0.321             | 0.185     |
| COAD   | G12V   | 0.195             | 0.196             | 1         |
| LUAD   | G12A   | 0.0892            | 0.0717            | 0.350     |
| LUAD   | G12D   | 0.122             | 0.139             | 0.552     |
| LUAD   | G12V   | 0.225             | 0.253             | 0.386     |
| MM     | G12R   | 0.0528            | 0.0621            | 0.612     |


---

Minor changes to make:

1. To waterfall plots in "90_03_mutation-burden-distribution": make shorter and wider
2. "20_35_rc-fisher-comparison/rc_mutations_distribition.svg": let each axis scale separately.