---
title: '2019-10-11'
author: ~
date: '2019-10-11'
slug: '2019-10-11'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2019-10-11'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---


## Analysis of hits from genetic interaction tests (cont.)

I began by fixing the UpSet plots such that they now include all sets and intersections.
(The updated images are shown in [yesterday's notebook]({{< ref "/post/2019-10-10-2019-10-10.md" >}}).)

## *A priori* gene lists

For the figure, I see the "cloud" (ie. high-level) network adjacent to the UpSet plots to describe that the comutation patterns are distinct between the alleles.
Then, the next two logical questions would be:

1. Are there any important/notable genes in there?
2. Are any specific functions enriched for the alleles?

I began this follow up with the first question.
As noted in [yesterday's notebook]({{< ref "/post/2019-10-10-2019-10-10.md" >}}), I have gathered the following gene lists:

- the following KEGG gene sets (from [MSigDB](http://software.broadinstitute.org/gsea/downloads.jsp)) (a full list is available at the end of this notebook): 
    - apoptosis,
    - cell-cycle
    - ERBB signaling pathway
    - Hedgehog signaling pathway
    - MAPK signaling pathway
    - mTOR signaling pathway
    - Notch signaling pathway
    - p53 signaling pathway
    - VEGF signaling pathway
    - WNT signaling pathway
- the interactors with WT *KRAS* from [Kovalski *et al.*, 2019](https://www.ncbi.nlm.nih.gov/pubmed/?term=30639242)
- and the [COSMIC Cancer Gene Consensus](https://cancer.sanger.ac.uk/census)

The analysis was conducted in "src/20_43_apriori-lists-genetic-interactions.R".

During the analysis, I had to go back to "munge/30_rc-test_results-prep.R" and "src/20_35_rc-fisher-comparison.R" to fix some filtering mistakes.
Their effects on the results were minor.

The resulting plots are shown below.
For some of the LUAD and PAAD, I present separate plots for each source of gene list to keep the plot legible.


### COAD

![](/img/graphs/20_43_apriori-lists-genetic-interactions/goi_overlap_genetic_interactions_network_COAD_allLists.svg)


### LUAD

**KEGG pathways**

![](/img/graphs/20_43_apriori-lists-genetic-interactions/goi_overlap_genetic_interactions_network_LUAD_kegg.svg)

**Physical Interactors**

![](/img/graphs/20_43_apriori-lists-genetic-interactions/goi_overlap_genetic_interactions_network_LUAD_BioID.svg)

**CGC Genes**

![](/img/graphs/20_43_apriori-lists-genetic-interactions/goi_overlap_genetic_interactions_network_LUAD_cgc.svg)


### MM

![](/img/graphs/20_43_apriori-lists-genetic-interactions/goi_overlap_genetic_interactions_network_MM_allLists.svg)


### PAAD

**KEGG pathways**

![](/img/graphs/20_43_apriori-lists-genetic-interactions/goi_overlap_genetic_interactions_network_PAAD_kegg.svg)

**Physical Interactors**

![](/img/graphs/20_43_apriori-lists-genetic-interactions/goi_overlap_genetic_interactions_network_PAAD_BioID.svg)

**CGC Genes**

![](/img/graphs/20_43_apriori-lists-genetic-interactions/goi_overlap_genetic_interactions_network_PAAD_cgc.svg)


### A few comments

- There are some genes that comutate with one allele and are mutually exclusive with another. This is something to look into.
- There are some obvious stand-outs in terms of how "strong" (low p-value) the interaction is. Again, these are likely worth further investigation.


## Functional Enrichment

Instead of using specific gene lists chosen for *a priori* reasons, I used a statistical tool to identify biological functions that are enriched in the list of genes identified to have a genetic interaction with a *KRAS* allele.

### ErncihR

The first attempt used the Enrichr [(Chen *et al.*, 2013)](https://www.ncbi.nlm.nih.gov/pubmed/23586463) tool available through a R API using the ['enrichR'](https://cran.r-project.org/web/packages/enrichR/index.html) package (link to the [web API](http://amp.pharm.mssm.edu/Enrichr/)).

The following data bases were used:

- "BioCarta_2016",
- "GO_Biological_Process_2018",
- "KEA_2015",
- "KEGG_2019_Human",
- "LINCS_L1000_Kinase_Perturbations_down",
- "LINCS_L1000_Kinase_Perturbations_up",
- "Panther_2016",
- "PPI_Hub_Proteins",
- "Reactome_2016",
- "Transcription_Factor_PPIs",
- "WikiPathways_2019_Human"

A complete list can be found using the following command in R.

```r
sort(enrichR::listEnrichrDbs()$libraryName)
```

The results were filtered by adjusted p-value < 0.20 and at least 2 genes in the gene set.
The filtered tables were written to "tables/20_45_fxnal-enrich-genetic-interactions".


**TODO**

There is a bug where not all of the interactions are being included.
For example, *BRAF* is only connected to G13D in the KEGG-specific interaction network.







---

### Full List of KEGG Genesets

GLYCOLYSIS_GLUCONEOGENESIS, CITRATE_CYCLE_TCA_CYCLE, PENTOSE_PHOSPHATE_PATHWAY, PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS, FRUCTOSE_AND_MANNOSE_METABOLISM, GALACTOSE_METABOLISM, ASCORBATE_AND_ALDARATE_METABOLISM, FATTY_ACID_METABOLISM, STEROID_BIOSYNTHESIS, PRIMARY_BILE_ACID_BIOSYNTHESIS, STEROID_HORMONE_BIOSYNTHESIS, OXIDATIVE_PHOSPHORYLATION, PURINE_METABOLISM, PYRIMIDINE_METABOLISM, ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM, GLYCINE_SERINE_AND_THREONINE_METABOLISM, CYSTEINE_AND_METHIONINE_METABOLISM, VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION, VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS, LYSINE_DEGRADATION, ARGININE_AND_PROLINE_METABOLISM, HISTIDINE_METABOLISM, TYROSINE_METABOLISM, PHENYLALANINE_METABOLISM, TRYPTOPHAN_METABOLISM, BETA_ALANINE_METABOLISM, TAURINE_AND_HYPOTAURINE_METABOLISM, SELENOAMINO_ACID_METABOLISM, GLUTATHIONE_METABOLISM, STARCH_AND_SUCROSE_METABOLISM, N_GLYCAN_BIOSYNTHESIS, OTHER_GLYCAN_DEGRADATION, O_GLYCAN_BIOSYNTHESIS, AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM, GLYCOSAMINOGLYCAN_DEGRADATION, GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE, GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE, GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE, GLYCEROLIPID_METABOLISM, INOSITOL_PHOSPHATE_METABOLISM, GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS, GLYCEROPHOSPHOLIPID_METABOLISM, ETHER_LIPID_METABOLISM, ARACHIDONIC_ACID_METABOLISM, LINOLEIC_ACID_METABOLISM, ALPHA_LINOLENIC_ACID_METABOLISM, SPHINGOLIPID_METABOLISM, GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES, GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES, GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES, PYRUVATE_METABOLISM, GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM, PROPANOATE_METABOLISM, BUTANOATE_METABOLISM, ONE_CARBON_POOL_BY_FOLATE, RIBOFLAVIN_METABOLISM, NICOTINATE_AND_NICOTINAMIDE_METABOLISM, PANTOTHENATE_AND_COA_BIOSYNTHESIS, FOLATE_BIOSYNTHESIS, RETINOL_METABOLISM, PORPHYRIN_AND_CHLOROPHYLL_METABOLISM, TERPENOID_BACKBONE_BIOSYNTHESIS, LIMONENE_AND_PINENE_DEGRADATION, NITROGEN_METABOLISM, SULFUR_METABOLISM, AMINOACYL_TRNA_BIOSYNTHESIS, METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450, DRUG_METABOLISM_CYTOCHROME_P450, DRUG_METABOLISM_OTHER_ENZYMES, BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS, ABC_TRANSPORTERS, RIBOSOME, RNA_DEGRADATION, RNA_POLYMERASE, BASAL_TRANSCRIPTION_FACTORS, DNA_REPLICATION, SPLICEOSOME, PROTEASOME, PROTEIN_EXPORT, PPAR_SIGNALING_PATHWAY, BASE_EXCISION_REPAIR, NUCLEOTIDE_EXCISION_REPAIR, MISMATCH_REPAIR, HOMOLOGOUS_RECOMBINATION, NON_HOMOLOGOUS_END_JOINING, MAPK_SIGNALING_PATHWAY, ERBB_SIGNALING_PATHWAY, CALCIUM_SIGNALING_PATHWAY, CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION, CHEMOKINE_SIGNALING_PATHWAY, PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM, NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION, CELL_CYCLE, OOCYTE_MEIOSIS, P53_SIGNALING_PATHWAY, UBIQUITIN_MEDIATED_PROTEOLYSIS, SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT, REGULATION_OF_AUTOPHAGY, LYSOSOME, ENDOCYTOSIS, PEROXISOME, MTOR_SIGNALING_PATHWAY, APOPTOSIS, CARDIAC_MUSCLE_CONTRACTION, VASCULAR_SMOOTH_MUSCLE_CONTRACTION, WNT_SIGNALING_PATHWAY, DORSO_VENTRAL_AXIS_FORMATION, NOTCH_SIGNALING_PATHWAY, HEDGEHOG_SIGNALING_PATHWAY, TGF_BETA_SIGNALING_PATHWAY, AXON_GUIDANCE, VEGF_SIGNALING_PATHWAY, FOCAL_ADHESION, ECM_RECEPTOR_INTERACTION, CELL_ADHESION_MOLECULES_CAMS, ADHERENS_JUNCTION, TIGHT_JUNCTION, GAP_JUNCTION, COMPLEMENT_AND_COAGULATION_CASCADES, ANTIGEN_PROCESSING_AND_PRESENTATION, RENIN_ANGIOTENSIN_SYSTEM, TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY, NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY, RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY, CYTOSOLIC_DNA_SENSING_PATHWAY, JAK_STAT_SIGNALING_PATHWAY, HEMATOPOIETIC_CELL_LINEAGE, NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY, T_CELL_RECEPTOR_SIGNALING_PATHWAY, B_CELL_RECEPTOR_SIGNALING_PATHWAY, FC_EPSILON_RI_SIGNALING_PATHWAY, FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS, LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION, INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION, CIRCADIAN_RHYTHM_MAMMAL, LONG_TERM_POTENTIATION, NEUROTROPHIN_SIGNALING_PATHWAY, LONG_TERM_DEPRESSION, OLFACTORY_TRANSDUCTION, TASTE_TRANSDUCTION, REGULATION_OF_ACTIN_CYTOSKELETON, INSULIN_SIGNALING_PATHWAY, GNRH_SIGNALING_PATHWAY, PROGESTERONE_MEDIATED_OOCYTE_MATURATION, MELANOGENESIS, ADIPOCYTOKINE_SIGNALING_PATHWAY, TYPE_II_DIABETES_MELLITUS, TYPE_I_DIABETES_MELLITUS, MATURITY_ONSET_DIABETES_OF_THE_YOUNG, ALDOSTERONE_REGULATED_SODIUM_REABSORPTION, VASOPRESSIN_REGULATED_WATER_REABSORPTION, PROXIMAL_TUBULE_BICARBONATE_RECLAMATION, ALZHEIMERS_DISEASE, PARKINSONS_DISEASE, AMYOTROPHIC_LATERAL_SCLEROSIS_ALS, HUNTINGTONS_DISEASE, PRION_DISEASES, VIBRIO_CHOLERAE_INFECTION, EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION, PATHOGENIC_ESCHERICHIA_COLI_INFECTION, LEISHMANIA_INFECTION, PATHWAYS_IN_CANCER, COLORECTAL_CANCER, RENAL_CELL_CARCINOMA, PANCREATIC_CANCER, ENDOMETRIAL_CANCER, GLIOMA, PROSTATE_CANCER, THYROID_CANCER, BASAL_CELL_CARCINOMA, MELANOMA, BLADDER_CANCER, CHRONIC_MYELOID_LEUKEMIA, ACUTE_MYELOID_LEUKEMIA, SMALL_CELL_LUNG_CANCER, NON_SMALL_CELL_LUNG_CANCER, ASTHMA, AUTOIMMUNE_THYROID_DISEASE, SYSTEMIC_LUPUS_ERYTHEMATOSUS, ALLOGRAFT_REJECTION, GRAFT_VERSUS_HOST_DISEASE, PRIMARY_IMMUNODEFICIENCY, HYPERTROPHIC_CARDIOMYOPATHY_HCM, ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC, DILATED_CARDIOMYOPATHY, VIRAL_MYOCARDITIS