---
title: '2019-10-07'
author: ~
date: '2019-10-07'
slug: '2019-10-07'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2019-10-07'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---


# RC Co-mutation test


## Introduction

A while back, I thought of using the RC-test for co-mutation and mutual exclusivity from [*A weighted exact test for mutually exclusive mutations in cancer* (Leiserson *et al.*, 2016)](https://academic.oup.com/bioinformatics/article/32/17/i736/2450792/) in place of the Fisher's exact test that Giorgio had employed.
The benefit of this test is that it accounts for the number of mutations *per sample* in its estimate of significance.
The main draw back is that it is permutation-based, and thus very slow.

I was able to write a [Snakemake](https://snakemake.readthedocs.io) script to run the test for all cancers and *KRAS* alleles. 
It has finally completed all of the tests and I will now port it over to this project.
To move forward with the analysis, I will just port over the results and begin parsing.
I will go back and move the files for running the analysis over in the future.
For now, it is located on O2 at the following URL: "/n/data2/bidmc/medicine/haigis/Cook/KrasComutation/FinalAnalyses/RC_comutation_2/"


## Munging

I copied the directory of the output of the RC-test into the data dir.

```bash
mkdir data/rc-test-results
cp -r ../KrasComutation/FinalAnalyses/RC_comutation_2/output/* data/rc-test-results
```

These files are automatically loaded into R by running `load.project()`.
I combined all of the information into a single data frame, `rc_test_results`, in the munging script "30_rc-test_results-prep.R".
Here, I also extracted the non-*KRAS* gene from the `gene_set` column.
The raw files were added to the list of raw data to ignore in "config/global.dcf".

## Analyzing RC-test results


### General scope of results

Below is the number of significant interactions (p-value < 0.05 and at-least 3 events) per cancer found by the RC-test.

![](/img/graphs/20_30_rc-test-results-analysis/sig_results_barplot.svg)

The plot below separates the bars by the type of interaction.

![](/img/graphs/20_30_rc-test-results-analysis/sig_results_barplot_septest.svg)


### Interaction networks

![](/img/graphs/20_30_rc-test-results-analysis/COAD_comut-mutex_network.svg)

![](/img/graphs/20_30_rc-test-results-analysis/LUAD_comut-mutex_network.svg)

![](/img/graphs/20_30_rc-test-results-analysis/PAAD_comut-mutex_network.svg)

![](/img/graphs/20_30_rc-test-results-analysis/MM_comut-mutex_network.svg)


## Moving the RC-test directory to this project

I got this far and decided that moving forward with any analysis would require a lot of the information from the RC-test directory.
Therefore, I stopped and ported it over.
The following are the steps I took.

### Moving input, intermediate, and output data

```bash
mkdir data/rc-test
mkdir data/rc-test/input  data/rc-test/intermediate data/rc-test/output
```


```bash
cp ../KrasComutation/FinalAnalyses/RC_comutation_2/input/* data/rc-test/input/
cp ../KrasComutation/FinalAnalyses/RC_comutation_2/intermediate/* data/rc-test/intermediate/
cp -r ../KrasComutation/FinalAnalyses/RC_comutation_2/output/* data/rc-test/output/
cp -r ../KrasComutation/FinalAnalyses/RC_comutation_2/permuted_graphs data/rc-test/
```

### Moving example DAG image

```bash
cp ../KrasComutation/FinalAnalyses/RC_comutation_2/images/dag.pdf graphs/20_20_rc-test_Snakefile/
```


### Moving scripts

```bash
cp ../KrasComutation/FinalAnalyses/RC_comutation_2/Snakefile src/20_20_rc-test_Snakefile
cp ../KrasComutation/FinalAnalyses/RC_comutation_2/snakemake_scripts/* src/
cp ../KrasComutation/FinalAnalyses/RC_comutation_2/prepare_input_tables.R src/
cp ../KrasComutation/FinalAnalyses/RC_comutation_2/generate_sm_output_names.R src
cp ../KrasComutation/FinalAnalyses/RC_comutation_2/cluster.json config/

```

The snakemake scripts were renamed.
I did my best to update the Snakefile for the new directory, but real testing will be required to truly find out.
A bash script was created to run the Snakemake file on O2, "20_21_run-rc-test-snakemake.sh".


## Munging cancer data from Giorgio Melloni

GM used a deeply-nested data structure for the cancer data.
His format is not conducive to my coding style (and is very large).
I created another munging script to prepare this data, "20_prepare-cancer-data.R".
The files from GM were copied from the old directory to "data/cancer-data/".
