---
title: '2019-09-28'
author: ~
date: '2019-09-28'
slug: '2019-09-28'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2019-09-28T15:28:05-04:00'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---

I am beginning the migration process from the only comutation directory to this project.

## Downloading DepMap 2019 Q3 release

The first step is moving over the raw data into the new data directory.
DepMap release 19Q3 data, so the new data sets were downloaded into "data/depmap19Q3/"

```bash
wget -O Achilles_gene_effect.csv https://ndownloader.figshare.com/files/16757666

wget -O Achilles_gene_dependency.csv https://ndownloader.figshare.com/files/16757660

wget -O Achilles_guide_efficacy.csv https://ndownloader.figshare.com/files/16757669

wget -O Achilles_common_essentials.csv https://ndownloader.figshare.com/files/16757654

wget -O common_essentials.csv https://ndownloader.figshare.com/files/16757714

wget -O nonessentials.csv https://ndownloader.figshare.com/files/16757717

wget -O Achilles_guide_map.csv https://ndownloader.figshare.com/files/16757672

wget -O CCLE_expression.csv https://ndownloader.figshare.com/files/16757690

wget -O CCLE_gene_cn.csv https://ndownloader.figshare.com/files/16757699

wget -O CCLE_fusions.csv https://ndownloader.figshare.com/files/16757696

wget -O CCLE_mutations.csv https://ndownloader.figshare.com/files/16757702

wget -O sample_info.csv https://ndownloader.figshare.com/files/16757723
```

## Munging DepMap 2019 Q3 release

I then created the munging script called "10_depmap19Q3.R" to prepare the data downloaded from DepMap.

### Essentiality

A single data frame was created for all of the data concerning whether a gene is labeled essential or non-essential.
To help, I created `get_huo_from_depmap_ids()` and `get_entrez_from_depmap_ids()` to extract the appropriate names from the data tables.
These functions are in "libs/helpers.R".
Tests were written in "tests/test-helpers.R".

### Copy Number

A data frame of CNA was created `ccle_copy_number`.
According to the README from DepMap, the copy number values were "Gene level copy number data, log2 transformed with a pseudo count of 1."
However, looking at the values when converted back ($2^{x}$), I do not think there is a pseudo count because most of the values are 1 is there was.
Therefore, I did not subtract 1 from the conversion.
I also added a label for homozygous deletion ("homo_del"), heterozygous deletion ("het_del"), amplification ("amp"), or no change ("norm").

### Mutations

The CCLE mutations were tidied up, though not much was needed.
I cached two data frames: `ccle_mutations` has all of the mutations, and `ccle_mutations_coding` has only coding mutations.

### RNA Expression

The RNA expression data is described by DepMap as follows:

> RNAseq TPM gene expression data for just protein coding genes using RSEM. Log2 transformed, using a pseudo-count of 1.

The tall formatted data was cached as `ccle_expression`.


### *KRAS* mutants

A tibble was made to hold the mutation information KRAS.
This included coding mutations and copy number amplifications.

## TODO

* Finish munging the new DepMap data.
