---
title: '2020-02-19'
author: ~
date: '2020-02-19'
slug: '2020-02-19'
categories: []
tags: []
subtitle: ''
summary: ''
authors: []
lastmod: '2020-02-19'
featured: no
image:
  caption: ''
  focal_point: ''
  preview_only: no
projects: []
---

## Calculate the frequency of the codons accounting for the prevalence of the cancers

In the paper I state the fraction of mutations at each codon, but I do not account for the different number of samples from each cancer.
After KH pointed this out, I also thought it would be a more useful metric if it were to account for the prevalence of the cancers.

### Sources

I collected cancer prevalence data from the following sources.
The data is downloaded in "./data/cancer-rates/".

I downloaded incidence data from 2012-2016 from SEER: [SEER Cancer Statistics Review 1975-2016](https://seer.cancer.gov/csr/1975_2016/browse_csr.php?sectionSEL=1&pageSEL=sect_01_table.04).
The data is in "table1-4_incidence-death-survival-rates-all-races.csv" and prepared into the data frame `seer_cancer_incidence_df`.

From The American Cancer Society, "In general, about 13% of all lung cancers are SCLC, and 84% are NSCLC" (["Key Statistics for Lung Cancer", *ACS*](https://www.cancer.org/cancer/lung-cancer/about/key-statistics.html)).
I added this value as a global constant in "./lib/cancer-specific_helpers.R" call `FRACTION_OF_LUNG_THAT_ARE_LUAD`.

I downloaded their estimated 2020 incidence data from [Cancer Statistics Center](https://cancerstatisticscenter.cancer.org/?_ga=2.9802422.205493334.1582111737-1270919539.1456557450#!/).
This is saved as "acs_NewCaseEstimates.xlsx".
I downloaded their 2012-2019 incidence rates (which they get from the North American Association of Central Cancer Registries (NAACCR)) to "acs_IncRate.xlsx".

### Calculation

I calculated the fraction of *KRAS* mutations at each codon adjusting for the incidence of the cancer.

I used the ACS numbers becasue I think their sampling procedure was more diverse than that used by SEER.
There were substantial differences in incidence between the two sources for all cancer and lung cancer.

![](img/graphs/90_05_kras-allele-distribution/cancer_incidence_stacked_plot.svg)
![](img/graphs/90_05_kras-allele-distribution/cancer_incidence_plot.svg)

I also adjusted the lung cancer figure to only account for the NSCLC (LUAD).

The "avg_codon_freq" is the average frequency of mutations at each codon in each cancer.
The "adj_codon_frequency" is the frequency of mutation at each codon weighted by the incidence of the cancer.

|codon | avg_codon_freq| adj_codon_freq|
|:-----|--------------:|--------------:|
|12    |      0.7220490|      0.7887384|
|13    |      0.0976307|      0.1052006|
|61    |      0.1483235|      0.0752111|
|146   |      0.0319968|      0.0308499|


## Survival analysis

KH was interested in looking at the survival curves for comutation genes as a part of his "alternative hypothesis."

### Data

As far as I could see, the only data sets with survival data over time were the TCGA data.
They all had overall survival (OS), progression free survival (PFS), and disease free (DF).
All of the data shown on GDC for COAD, LUAD, and PAAD are in cBioPortal, so I obtained the data from there.
I downloaded clinical data on MM from MMRF's Research Gateway.

The data was prepared in "./munge/70_survival-data.R".

### Tutorials

* ["Survival Analysis with R"](https://rviews.rstudio.com/2017/09/25/survival-analysis-with-r/) by R Views
* ["Survival Analysis in R"](https://www.emilyzabor.com/tutorials/survival_analysis_in_r_tutorial.html) by Emily C. Zabor (MSK)
* ["Survival Analysis Basics"](http://www.sthda.com/english/wiki/survival-analysis-basics) by STHDA
* ["Survival Analysis in R For Beginners"](https://www.datacamp.com/community/tutorials/survival-analysis-R) by DataCamp
* ['survival' R package vignette](https://cran.r-project.org/web/packages/survival/vignettes/survival.pdf) by Terry Therneau
